library/std/src/sync/reentrant_lock.rs

```diff
@@ -72,14 +73,95 @@ use crate::sys::sync as sys;
 // since we're not dealing with multiple threads. If it's not equal,
 // synchronization is left to the mutex, making relaxed memory ordering for
 // the `owner` field fine in all cases.
+//
+// On systems without 64 bit atomics we use a simple seqlock to emulate a 64 bit Tid using
+// 32 bit atomics (which should be supported on all platforms with `std`). This works
+// because only one thread at a time (the one holding the mutex) writes to it.
 #[unstable(feature = "reentrant_lock", issue = "121440")]
 pub struct ReentrantLock<T: ?Sized> {
     mutex: sys::Mutex,
-    owner: AtomicUsize,
+    owner: Tid,
     lock_count: UnsafeCell<u32>,
     data: T,
 }
 
+cfg_if!(
+    if #[cfg(target_has_atomic = "64")] {
+        use crate::sync::atomic::{AtomicU64, Ordering::Relaxed};
+
+        struct Tid(AtomicU64);
+
+        impl Tid {
+            const fn new(tid: u64) -> Self {
+                Self(AtomicU64::new(tid))
+            }
+
+            #[inline]
+            fn get(&self) -> u64 {
+                self.0.load(Relaxed)
+            }
+
+            #[inline]
+            fn set(&self, tid: u64) {
+                self.0.store(tid, Relaxed)
+            }
+        }
+    } else if #[cfg(target_has_atomic = "32")] {
+        use crate::sync::atomic::{AtomicU32, Ordering::{Acquire, Relaxed, Release}};
+
+        struct Tid {
+            seq: AtomicU32,
+            low: AtomicU32,
+            high: AtomicU32,
+        }
+
+        impl Tid {
+            const fn new(tid: u64) -> Self {
+                Self {
+                    seq: AtomicU32::new(0),
+                    low: AtomicU32::new(tid as u32),
+                    high: AtomicU32::new((tid >> 32) as u32),
+                }
+            }
+
+            #[inline]
+            fn get(&self) -> u64 {
+                // Synchronizes with the release-increment in `set()` to ensure
+                // we only read the data after it's been fully written.
+                let mut seq = self.seq.load(Acquire);
+                loop {
+                    if seq % 2 == 0 {
+                        let low = self.low.load(Relaxed);
+                        let high = self.high.load(Relaxed);
+                        // The acquire-increment in `set()` synchronizes with this release
+                        // store to ensure that `get()` doesn't see data from a subsequent
+                        // `set()` call.
+                        match self.seq.compare_exchange_weak(seq, seq, Release, Acquire) {
+                            Ok(_) => return u64::from(low) | (u64::from(high) << 32),
+                            Err(new) => seq = new,
+                        }
+                    } else {
+                        crate::hint::spin_loop();
+                        seq = self.seq.load(Acquire);
```

Comment:
No need to spin here: if there is a write in progress, the current thread cannot be the lock owner.